cpp/src/arrow/flight/client.cc

```diff
@@ -227,7 +77,9 @@ class FlightStreamReader : public RecordBatchReader {
     // For customizing read path for better memory/serialization efficiency
     auto custom_reader = reinterpret_cast<grpc::ClientReader<FlightData>*>(stream_.get());
 
-    if (custom_reader->Read(&data)) {
+    // Explicitly specify the override to invoke - otherwise compiler
+    // may invoke through vtable (not updated by reinterpret_cast)
```

Comment:
That's interesting. Which compiler did you observe this on? I didn't think this was a virtual method

Reply:
```
$ clang++ --version
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

It's inconsistent - the calls you originally had were done statically, but when I added another instance for DoPut, Clang opted for a virtual call, judging from the disassembly.

Reply:
Unfortunately this is a correctness issue; the reader is a generic class instantiated for the Protobuf-generated struct, and uses the Protobuf serializer/deserializer. We'd like to force it to instead user our serializer. The original implementation relied on using `reinterpret_cast` to change the template parameter, and then hoping the compiler generated the correct static call. I found that the compiler did not always do this, and would fall back to the Protobuf serialization path. This change is intended to force it not to do this.

This seems like undefined behavior to me, and it might be worthwhile to find a better way, but comments by Wes in the code indicate that gRPC doesn't expose a good way to do this.

Reply:
Hmm... What's the difference between `pb::FlightData` and `FlightData`?

Reply:
In any case, if `pb::FlightData` and `FlightData` are distinct classes, then `grpc::ClientReader<FlightData>` and `grpc::ClientReader<pb::FlightData>` are unrelated: how can one class call through the vtable of the other?

Reply:
Or if that happens, it means the `reinterpret_cast` is just illegal :-(

Reply:
Yes, that's why I think it's undefined behavior - it's not a legal `reinterpret_cast`.

Reply:
`pb::FlightData` is the Protobuf-generated struct, `FlightData` is our own struct (in `serialization-internal.h`)

Reply:
It would be nice to add a comment warning about the undefined behavior - so that this doesn't get lost then rediscovered.

Reply:
I've added a comment before the four uses of `reinterpret_cast`.